engine/src/main/java/org/hibernate/validator/internal/engine/validationcontext/ValidationContext.java

```diff
@@ -8,70 +8,74 @@
-import org.hibernate.validator.internal.metadata.core.MetaConstraint;
 import org.hibernate.validator.internal.metadata.descriptor.ConstraintDescriptorImpl;
 
 /**
- * Interface that expose contextual information required for a validation call.
+ * Context object interface keeping track of all required data for operations inside {@link ConstraintTree}
+ * and its subclasses.
  * <p>
- * Provides ability to collect failing constrains, as well as gives access to resources like
- * constraint validator factory, traversable resolver, etc.
+ * Allows to collect all failing constraints, creates {@link ConstraintValidatorContext}s based on the constraint
+ * descriptors, and exposes other resources needed to initialize a new {@link ConstraintValidator}.
  *
  * @author Hardy Ferentschik
  * @author Emmanuel Bernard
  * @author Gunnar Morling
  * @author Guillaume Smet
+ * @author Marko Bekhta
  */
 public interface ValidationContext<T> {
 
-	T getRootBean();
-
-	Class<T> getRootBeanClass();
-
-	BeanMetaData<T> getRootBeanMetaData();
-
-	TraversableResolver getTraversableResolver();
-
+	/**
+	 * @return {@code true} if the fail fast mode is enabled, {@code false} otherwise.
+	 */
 	boolean isFailFastModeEnabled();
 
+	/**
+	 * @return the {@link ConstraintValidatorManager} that caches and manages
+	 * 		life cycle of constraint validator instances.
+	 */
 	ConstraintValidatorManager getConstraintValidatorManager();
 
+	/**
+	 * @return constraint validator initialization context.
+	 */
 	HibernateConstraintValidatorInitializationContext getConstraintValidatorInitializationContext();
 
+	/**
+	 * @return constraint validator factory which should be used in this context.
+	 */
 	ConstraintValidatorFactory getConstraintValidatorFactory();
 
-	boolean isBeanAlreadyValidated(Object value, Class<?> group, PathImpl path);
-
-	void markCurrentBeanAsProcessed(ValueContext<?, ?> valueContext);
-
-	Set<ConstraintViolation<T>> getFailingConstraints();
-
+	/**
+	 * Creates a constraint violation {@link ConstraintViolation} based on passed parameters and adds it
+	 * to a set of failed constrains.
+	 */
 	void addConstraintFailure(
-			ValueContext<?, ?> localContext,
+			ValueContext<?, ?> valueContext,
 			ConstraintViolationCreationContext constraintViolationCreationContext,
 			ConstraintDescriptor<?> descriptor
 	);
 
-	boolean hasMetaConstraintBeenProcessed(Object bean, Path path, MetaConstraint<?> metaConstraint);
-
-	void markConstraintProcessed(Object bean, Path path, MetaConstraint<?> metaConstraint);
-
-	default boolean canProcess(MetaConstraint<?> metaConstraint) {
-		return true;
-	}
+	/**
+	 * @return a set of constraint vilations created with {@link ValidationContext#addConstraintFailure(ValueContext, ConstraintViolationCreationContext, ConstraintDescriptor)}.
+	 */
+	Set<ConstraintViolation<T>> getFailingConstraints();
 
+	/**
+	 * @return created constraint validator context based on passed constraint descriptor and path.
```

Comment:
It would be more readable as `a newly created constraint validator context based on the passed constraint descriptor and path.`